GowerMetric.py

```diff
@@ -0,0 +1,371 @@
+
+            if self.weights is not None:
+                ratio_dist = ratio_dist @ self.weights[self._ratio_scale_idx]
+            else:
+                ratio_dist = ratio_dist.sum()
+        else:
+            ratio_dist = 0.0
+
+        # Sum of all distances
+        distance = cat_nom_dist + ratio_dist
+
+        # Normalization
+        distance /= self.n_features_in_
+
+        return distance
+
+    def fit(self, X):
+        self.ranges_ = np.zeros(self.dtypes.size)
+        self.h_ = np.zeros(self.dtypes.size)
+        self.n_features_in_ = self.dtypes.size
+
+        for i in range(self.dtypes.size):
+            if self.dtypes[i] == DataType.RATIO_SCALE:
+                column = X[:, i]
+
+                if self.ratio_scale_normalization == "iqr":
+                    # IQR (g_t) - Interquartile Range
+                    q1, q3 = np.percentile(column, [25, 75])
+                    self.ranges_[i] = q3 - q1
+
+                    if self.ranges_[i] == 0:
+                        self.ranges_[i] = np.ptp(column, axis=0)
+
+                elif self.ratio_scale_normalization == "range":
+                    # Traditional range of values
+                    self.ranges_[i] = np.ptp(column, axis=0)
+                else:
+                    print(
+                        "GowerMetric - Ratio scale has wrong type of normalization!"
+                    )
+                    raise ValueError
+
+                # h_t - bandwidth in the kernel density estimation (Marcello D’Orazio - p. 9)
+                c = 1.06
+
+                s = np.std(column)
+                n = X[:, i].size
+                self.h_[i] = (
+                    c / n ** (1 / 5) * np.min([s, self.ranges_[i] / 1.34])
+                )
+
+        if self.weights == "precomputed":
+            loader = GowerMetricWeightsLoader(self)
+            loader.load_weights(self.precomputed_weights_file)
+        elif self.weights == "cpcc":
+            loader = GowerMetricWeightsLoader(self)
+            loader.select_weights(X)
+
+        self._select_number_of_clusters(X)
+
+    def cat_nom_bin_sym(
+        self, vector_1: np.ndarray, vector_2: np.ndarray
+    ) -> np.ndarray:
+
+        return vector_1 == vector_2
+
+    def ratio_scale(
+        self,
+        vector_1: np.ndarray,
+        vector_2: np.ndarray,
+        val_range: np.ndarray,
+        h: np.ndarray,
+    ) -> np.ndarray:
+
+        absolute = vector_1 - vector_2
+        absolute = np.abs(absolute)
+
+        ones = absolute >= val_range
+        zeros = absolute <= h
+
+        if vector_1.size == 1:
+            if ones:
+                return np.array(1.0)
+            elif zeros:
+                return np.array(0.0)
+            else:
+                absolute /= val_range
+        else:
+            absolute /= val_range
+            absolute[zeros] = 0.0
+            absolute[ones] = 1.0
+
+        return 1.0 - absolute
+
+    def _select_number_of_clusters(self, X):
```

Comment:
Move everything for calculating weights to a separate file

Reply:
I do not review weighting for now, as this is out of score for the current "sprint"